### PR TITLE
bpo-40116: Add insertion order bit-vector to dict values to allow dicts to share keys more freely.

### PR DESCRIPTION
**Pull request by markshannon**
See https://github.com/faster-cpython/ideas/issues/30#issuecomment-923960490

https://bugs.python.org/issue40116

### CODE REVIEW

**Comment by markshannon** [comment]
No significant performance change https://gist.github.com/markshannon/814e38650cbe4ba1639e31294d4879e6

File: Objects/dictobject.c [hunk 1]
```diff
@@ -458,6 +460,12 @@ static PyObject *empty_values[1] = { NULL };
 #  define ASSERT_CONSISTENT(op) assert(_PyDict_CheckConsistency((PyObject *)(op), 0))
 #endif
 
+static inline int
+get_index_from_order(PyDictObject *mp, int i)
+{
+    int shift = (mp->ma_used-1-i)*4;
```
Comment (methane) [inline Objects/dictobject.c]: Would you fix these warnings from Windows?

File: Objects/dictobject.c [hunk 2]
```diff
@@ -1468,6 +1502,9 @@ _PyDict_GetItemStringWithError(PyObject *v, const char *key)
 /* Fast version of global value lookup (LOAD_GLOBAL).
  * Lookup in globals, then builtins.
  *
+ *
+ *
+ *
```
Comment (methane) [inline Objects/dictobject.c]: remove this.

File: Objects/dictobject.c [hunk 3]
```diff
@@ -1173,13 +1188,14 @@ dictresize(PyDictObject *mp, uint8_t log2_newsize)
          * Note that values of split table is always dense.
          */
         for (Py_ssize_t i = 0; i < numentries; i++) {
-            assert(oldvalues[i] != NULL);
-            PyDictKeyEntry *ep = &oldentries[i];
+            int index = oldvalues->mv_order >> ((numentries-1-i)*4)&15;
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
            int index = (oldvalues->mv_order >> ((numentries-1-i)*4)) & 15;
```

File: Objects/dictobject.c [hunk 4]
```diff
@@ -1569,11 +1606,30 @@ delitem_common(PyDictObject *mp, Py_hash_t hash, Py_ssize_t ix,
...
+        assert(old_value == mp->ma_values->values[ix]);
+        mp->ma_values->values[ix] = NULL;
+        assert(ix < SHARED_KEYS_MAX_SIZE);
+        /* Update order */
+        for (int i = 0;; i+= 4) {
+            assert (i < 64);
+            if (((mp->ma_values->mv_order >> i) & 15) == (uint64_t)ix) {
+                /* Remove 4 bits at ith position */
+                uint64_t order = mp->ma_values->mv_order;
+                uint64_t high = order >> i >> 4 << i;
```
Comment (methane) [inline Objects/dictobject.c]: Please use parentheses here.

**Comment by markshannon** [comment]
Despite providing no speedup, I'd like to merge this, as it enables [compact objects](https://github.com/faster-cpython/ideas/issues/72).
Work-in-progress implementation of compact objects here:  https://github.com/faster-cpython/cpython/tree/inline-attributes-with-opt (there are a few leaks that I need to hunt down, and some general cleanup to be done).

**Comment by markshannon** [comment]
The build bot failures are cancellations (presumably timeouts).

File: Objects/dictobject.c [hunk 5]
```diff
@@ -2534,7 +2549,7 @@ dict_merge(PyObject *a, PyObject *b, int override)
             key = entry->me_key;
             hash = entry->me_hash;
             if (other->ma_values)
-                value = other->ma_values[i];
+                value = other->ma_values->values[i];
```
Comment (methane) [inline Objects/dictobject.c:2552]: I noticed that this change may break order of merged dict.

**Pull request merged**